test/extended/router/http2.go

```diff
@@ -0,0 +1,217 @@
+			}, {
+				routeType:         routev1.TLSTerminationReencrypt,
+				routePrefix:       "http2-default-cert",
+				useHTTP2Transport: true,
+				expectedGetError:  `http2: unexpected ALPN protocol ""; want "h2"`,
+			}, {
+				routeType:         routev1.TLSTerminationPassthrough,
+				routePrefix:       "http2-default-cert",
+				frontendProto:     "HTTP/2.0",
+				backendProto:      "HTTP/2.0",
+				statusCode:        http.StatusOK,
+				useHTTP2Transport: true,
+			}, {
+				routeType:         routev1.TLSTerminationEdge,
+				routePrefix:       "http2-custom-cert",
+				frontendProto:     "HTTP/1.1",
+				backendProto:      "HTTP/1.1",
+				statusCode:        http.StatusOK,
+				useHTTP2Transport: false,
+			}, {
+				routeType:         routev1.TLSTerminationReencrypt,
+				routePrefix:       "http2-custom-cert",
+				frontendProto:     "HTTP/1.1",
+				backendProto:      "HTTP/2.0", // reencrypt always has backend ALPN enabled
+				statusCode:        http.StatusOK,
+				useHTTP2Transport: false,
+			}, {
+				routeType:         routev1.TLSTerminationPassthrough,
+				routePrefix:       "http2-custom-cert",
+				frontendProto:     "HTTP/1.1",
+				backendProto:      "HTTP/1.1",
+				statusCode:        http.StatusOK,
+				useHTTP2Transport: false,
+			}, {
+				routeType:         routev1.TLSTerminationEdge,
+				routePrefix:       "http2-default-cert",
+				frontendProto:     "HTTP/1.1",
+				backendProto:      "HTTP/1.1",
+				statusCode:        http.StatusOK,
+				useHTTP2Transport: false,
+			}, {
+				routeType:         routev1.TLSTerminationReencrypt,
+				routePrefix:       "http2-default-cert",
+				frontendProto:     "HTTP/1.1",
+				backendProto:      "HTTP/2.0", // reencrypt always has backend ALPN enabled
+				statusCode:        http.StatusOK,
+				useHTTP2Transport: false,
+			}, {
+				routeType:         routev1.TLSTerminationPassthrough,
+				routePrefix:       "http2-default-cert",
+				frontendProto:     "HTTP/1.1",
+				backendProto:      "HTTP/1.1",
+				statusCode:        http.StatusOK,
+				useHTTP2Transport: false,
+			}}
+
+			for i, tc := range testCases {
+				testConfig := fmt.Sprintf("%+v", tc)
+				e2e.Logf("[test #%d/%d]: config: %s", i+1, len(testCases), testConfig)
+
+				// check readiness probe is accessible
+				urlTester := url.NewTester(oc.AdminKubeClient(), oc.KubeFramework().Namespace.Name).WithErrorPassthrough(true)
+				defer urlTester.Close()
+				hostname := getHostnameForRoute(oc, fmt.Sprintf("%s-%s", tc.routePrefix, tc.routeType))
+				urlTester.Within(30*time.Second, url.Expect("GET", "https://"+hostname+"/healthz").Through(hostname).SkipTLSVerification().HasStatusCode(200))
+
+				var resp *http.Response
+				client := makeHTTPClient(tc.useHTTP2Transport, http2ClientTimeout)
+
+				o.Expect(wait.Poll(time.Second, 30*time.Second, func() (bool, error) {
+					var err error
+					url := "https://" + hostname
+					resp, err = client.Get(url)
+					if err != nil && tc.expectedGetError != "" {
```

Comment:
Better to use `len(tc.expectedGetError) != 0`.

Reply:
oh, this is a kube-ism. :)